operators/src/opencl/cl_program.rs

```diff
@@ -0,0 +1,1743 @@
+
+                let nulls_buffer = if let Some(nulls_buffer) = column_buffer.nulls {
+                    // TODO: read i32 into null buffers as bool more efficiently
+                    let mut nulls = vec![0_i32; builder.num_features()];
+                    nulls_buffer.read(&mut nulls).enq()?;
+
+                    let num_bytes = bit_util::ceil(nulls.len(), 8);
+                    let mut arrow_buffer =
+                        MutableBuffer::new(num_bytes).with_bitset(num_bytes, false);
+                    let null_slice = arrow_buffer.data_mut();
+
+                    for (i, null) in nulls.iter().enumerate() {
+                        if *null != 0 {
+                            bit_util::set_bit(null_slice, i);
+                        }
+                    }
+                    Some(arrow_buffer.freeze())
+                } else {
+                    None
+                };
+
+                builder.set_column::<Int64Type>(
+                    &column_buffer.column_name,
+                    values_buffer,
+                    nulls_buffer,
+                )?;
+            }
+
+            // TODO: string, category columns
+
+            // TODO: time
+            builder.set_default_time_intervals()?;
+
+            builder.finish()?;
+        }
+        Ok(())
+    }
+
+    fn read_ocl_to_arrow_buffer<T: OclPrm>(
+        ocl_buffer: &Buffer<T>,
+        len: usize,
+    ) -> Result<arrow::buffer::Buffer> {
+        let mut arrow_buffer = MutableBuffer::new(len * std::mem::size_of::<T>());
+        arrow_buffer.resize(len * std::mem::size_of::<T>()).unwrap();
+
+        let dest = unsafe {
+            std::slice::from_raw_parts_mut(arrow_buffer.data_mut().as_ptr() as *mut T, len)
+        };
+
+        ocl_buffer.read(dest).enq()?;
+
+        Ok(arrow_buffer.freeze())
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Default, PartialEq)]
+struct RasterInfo {
+    pub size: [cl_uint; 3],
+    pub origin: [cl_double; 3],
+    pub scale: [cl_double; 3],
+
+    pub min: cl_double,
+    pub max: cl_double,
+    pub no_data: cl_double,
+
+    pub crs_code: cl_ushort,
+    pub has_no_data: cl_ushort,
+}
+
+unsafe impl Send for RasterInfo {}
+unsafe impl Sync for RasterInfo {}
+unsafe impl OclPrm for RasterInfo {}
+
+impl RasterInfo {
+    pub fn from_raster<T: Pixel>(raster: &Raster2D<T>) -> Self {
+        // TODO: extract missing information from raster
+        Self {
+            size: [
+                raster.dimension().size_of_x_axis().as_(),
+                raster.dimension().size_of_y_axis().as_(),
+                1, // TODO
+            ],
+            origin: [0., 0., 0.],
+            scale: [0., 0., 0.],
+            min: 0.,
+            max: 0.,
+            no_data: raster.no_data_value.map_or(0., AsPrimitive::as_),
+            crs_code: 0,
+            has_no_data: u16::from(raster.no_data_value.is_some()),
```

Comment:
the u16 is what we used in mapping. You are right we should'nt mix types. Which one should I use?

Reply:
Unfortunately, the size of a boolean is not defined in the OpenCL spec, this is also why it can't be transferred to the device. So we can't just use the same size here.

There are other approaches:
* What is the size of bool on our devices?
* What aligns better with the memory we use, e.g., if we use only 32 bit types…?
* Is it better to use the smallest type to keep the transfer low? Then we should use 1 byte.

Reply:
yeah I really have no idea why you would use u16. I agree either use smallest (u8) or best suited for access, but I am not sure which that is. Should I run some benchmarks?

Reply:
It would be nice, otherwise, just use a type from an educated guess.

Reply:
I use i8 consistently now d11f857f76f5c874662f622748dea347aaa954a4